### ISSUE TITLE
IntelliSense not working on object instances imported from external module

### ISSUE THREAD

**Issue by rohit901** [microsoft/pylance-release#1152]
## Environment data

-   Language Server version: 2021.4.0 (pyright 2918da1d)
-   OS and version: Manjaro Linux, Kernel 5.10.26-1-MANJARO
-   Python version (& distribution if applicable, e.g. Anaconda): Python 3.8.5 (default, Sep  4 2020, 07:30:14) 
[GCC 7.3.0] :: Anaconda, Inc. on linux

## Expected behaviour

1. pip install selenium
2. use the code below
```
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC

PATH = "PATH_TO/chromedriver"
driver = webdriver.Chrome(PATH)

driver.get("https://enter-url.com/")
main = WebDriverWait(driver, 10).until(
        EC.presence_of_element_located((By.ID, "main"))
    )
x = main.f
```
I expect IntelliSense to give me suggestions for the methods of the variable main since main is an object instance in this external module selenium, but I do not get any suggestions for it in auto-complete, while suggestions work fine for "EC" In the above code.

## Actual behaviour

It should suggest me the list of available methods/instance variable for the instance main in the above code but I'm getting no suggestions even when I press ctrl + space

## Logs

```
XXX
```

## Code Snippet / Additional information

```python
XXX
```

**Comment by judej** [comment]
Thanks for the report. What type do you see when you hover over `main` in the assignment statement? Also, if you could hover over `until()` and see what type it returns, that will help us get to the root of the problem.

**Comment by rohit901** [comment]
When I hover over main, it shows: **(variable) main: Any**
Refer to the  image below to know the output when hovering over until()

![code_intelliSense](https://user-images.githubusercontent.com/30185369/114600842-6f477d80-9cb2-11eb-8fe7-33547958b7a8.png)

**Comment by jakebailey** [comment]
Yeah, this means that the library doesn't have good type info, and we're just saying it's "Any", hence no completion.

If you go-to-def on `until`, do you see a type annotation there?

**Comment by rohit901** [comment]
I think there is no type annotation here? refer to the screenshot attached.

![untilType](https://user-images.githubusercontent.com/30185369/114811708-40bdc580-9dcc-11eb-9290-0f9eb9c26559.png)

I mean when I tried to check for IntelliSense using Sublime editor it was able to show me some suggestions for methods of "main" but there is no suggestion at all in VS Code. 

Also, please refer to [13:25](https://youtu.be/b5jt2bhSeXs?list=PLzMcBGfZo4-n40rB1XaJ0ak1bemvlqumQ&t=806) of this video, here, the author is able to get suggestions of methods from main.

**Comment by erictraut** [comment]
It looks like type stubs for selenium were previously included in typeshed, but they were [removed](https://github.com/python/typeshed/pull/2137) because they were low quality and caused more problems than they solved. 

In this particular example, pylance attempts to use its most sophisticated type inference mechanism called "call-site inference" to analyze called functions using the argument types passed into it. Pylance does this up to three call levels deep before it gives up. It is computationally very expensive (exponential with the number of levels), so it's not feasible to analyze further. The correct solution here is for the library to provide definitive type information in the form of type annotations.

If the selenium library is important to you, you should encourage its maintainers to include type information (either packaged stubs or inlined type annotations). We have documented [this guidance](https://github.com/microsoft/pyright/blob/master/docs/typed-libraries.md) for library authors. Without type information, pylance will do the best it can to infer types, but there are limits to what it can do.

**Comment by rohit901** [comment]
oh, I see, thank you for helping me understand the reason somewhat. But I wanted to ask that how is Sublime working on giving correct suggestions as shown in the video [here](https://youtu.be/b5jt2bhSeXs?list=PLzMcBGfZo4-n40rB1XaJ0ak1bemvlqumQ&t=806), and also on my local machine. I mean I feel it doesn't take that much time for suggestions from Sublime on my machine too.

**Comment by erictraut** [comment]
I don't know what language server was being used with Sublime in that video or whether the person in the video had type stubs installed, so I can't answer your question. My guess is that he had installed (or created) type stubs for selenium. I've searched for publicly-available stubs, but I wasn't able to find any.

**Comment by rohit901** [comment]
I'm really sorry, but actually, I don't have much idea about type stubs. But can VS Code get the type stubs that are being used by Sublime Text 3? 
I mean, refer to this screenshot from my sublime text 3 IDE, as we can see that the suggestions are working fine.
![subLimeExample](https://user-images.githubusercontent.com/30185369/114823504-41ad2200-9de1-11eb-8169-88ffa90b97ba.png)

I'm using [Anaconda Package](http://damnwidget.github.io/anaconda/) for sublime in case you are wondering,

**Comment by jakebailey** [comment]
I don't know much about Sublime, but it's very possible it's using MPLS or Jedi to get this information.

**Comment by rohit901** [comment]
Can pylance also use the same logic/implementation then? I mean I prefer working in VS Code, and maybe this is not the only library that has this problem right? would be really nice if IntelliSense works for more cases.

**Comment by jakebailey** [comment]
MPLS and Jedi are completely different analysis engines; you can certainly try one of them in VS Code to see if the behavior is the same (change the langaugeServer setting), but this won't fix it for Pylance.

I'm guessing it's not MPLS (we never shipped selenium stubs there), so must be Jedi, which has its own analysis quirks, and probably goes and follows things more deeply than our limits (which we have in place for a good reason; perf and diminishing returns).
